library/board.py

```diff
@@ -19,40 +19,79 @@ def __init__(self) -> None:
+        subgrid = self.collect_subgrid(number)
+        return self.check_grid_victory(subgrid)
+
+    @staticmethod
+    def check_grid_victory(grid: npt.NDArray[np.str_]) -> typing.Optional[str]:
+        """Determine if a given 3x3 grid has been won, and by who."""
+        for player in ("X", "O"):
+            # this turns our grid of ["X", "O", "•"] into [True, False, False] by applying the condition to every spot
+            ticks = grid == player
+
+            # Summing over booleans treats True as 1, False as 0. 3 Trues = 3, so "won" rows sum to 3.
+            # Because this is a 3x3 board, this results in a list of 3 counts that can be compared to 3
+            # to determine True/False indicating whether each row was won by this player
+            horizontals = ticks.sum(axis=0) == 3
+            if horizontals.any():
+                return player
+            verticals = ticks.sum(axis=1) == 3
+            if verticals.any():
+                return player
+
+            # .diagonal() gets us a single diagonal list,
+            # so these two rows are just True/False whether a diagonal is won
+            diagonal_r = ticks.diagonal().sum() == 3
+            diagonal_l = ticks.fliplr().diagonal().sum() == 3
+            if diagonal_l or diagonal_r:
+                return player
+
+        return None
+
     def collect_subgrid(self, number: str) -> npt.NDArray[np.str_]:
         """
         Takes a subgrid choice and returns that np.ndarray and the choice str
 
         The subgrids are laid out as so:
-            6 7 8
-            3 4 5
-            0 1 2
+            7 8 9
+            4 5 6
+            1 2 3
         """
-        subgrid = np.full((3, 3), "·")
-        if number == "0":
-            subgrid = self.contents[6:, :3]
-        elif number == "1":
-            subgrid = self.contents[6:, 3:6]
-        elif number == "2":
-            subgrid = self.contents[6:, 6:]
-        elif number == "3":
-            subgrid = self.contents[:3, :3]
-        elif number == "4":
-            subgrid = self.contents[:3, 3:6]
-        elif number == "5":
-            subgrid = self.contents[:3, 6:]
-        elif number == "6":
-            subgrid = self.contents[3:6, :3]
-        elif number == "7":
-            subgrid = self.contents[3:6, 3:6]
-        elif number == "8":
-            subgrid = self.contents[3:6, 6:]
-
-        return subgrid
+        first = slice(3)
+        second = slice(3, 6)
+        third = slice(6, 9)
+        subgrid_map = {
+            "1": (third, first),
+            "2": (third, second),
+            "3": (third, third),
+            "4": (second, first),
+            "5": (second, second),
+            "6": (second, third),
+            "7": (first, first),
+            "8": (first, second),
+            "9": (first, third),
+        }
+
+        # Mypy cannot penetrate the slicing behavior here, so we assert that it is correct
```

Comment:
Not sure "penetrate" is the right word here... Something like "figure out" might work?

Actually, looking into this:

`npt.NDArray` is an alias for `np.ndarray[Any, np.dtype[ScalarType]]`

This means: `npt.NDArray[np.str_]` == `np.ndarray[Any, np.dtype[np.str_]]`... And `__getitem__` for that seems to be a TODO: https://github.com/numpy/numpy/blob/89c80ba606f4346f8df2a31cfcc0e967045a68ed/numpy/__init__.pyi#L1202-L1203

Maybe rewrite the comment to contain that information?